src/tribler/core/utilities/tribler_process_manager.py

```diff
@@ -0,0 +1,175 @@
+    global global_process_manager  # pylint: disable=global-statement
+    with _lock:
+        global_process_manager = process_manager
+
+
+def get_global_process_manager() -> Optional[ProcessManager]:
+    with _lock:
+        return global_process_manager
+
+
+def set_error(error_msg: Optional[str] = None, error_info: Optional[dict] = None,
+              exc: Optional[Exception] = None, replace: bool = False):
+    process_manager = get_global_process_manager()
+    if process_manager is None:
+        logger.warning('Cannot set error for process locker: no process locker global instance is set')
+    else:
+        process_manager.current_process.set_error(error_msg, error_info, exc, replace)
+        process_manager.save(process_manager.current_process)
+
+
+def with_retry(func):
+    @wraps(func)
+    def new_func(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except sqlite3.Error as e:
+            logger.warning(f'Retrying after the error: {e.__class__.__name__}: {e}')
+            return func(*args, **kwargs)
+    new_func: func
+    return new_func
+
+
+class ProcessManager:
+    def __init__(self, root_dir: Path, process_kind: ProcessKind, creator_pid: Optional[int] = None, **other_params):
+        self.root_dir = root_dir
+        self.filename = self._get_file_name(root_dir)
+        self.current_process = TriblerProcess.current_process(process_kind, creator_pid, **other_params)
+        self.primary_process = self.atomic_get_primary_process(process_kind, self.current_process)
+
+    @classmethod
+    def _get_file_name(cls, root_dir: Path) -> Path:  # The method is added for easier testing
+        return root_dir / DB_FILENAME
+
+    def connect(self) -> sqlite3.Connection:
+        connection = sqlite3.connect(str(self.filename))
+        try:
+            connection.execute('BEGIN EXCLUSIVE TRANSACTION')
+            connection.execute(CREATE_SQL)
+            return connection
+        except:  # noqa: E722
+            connection.close()
+            raise
+
+    @contextmanager
+    def transaction(self) -> ContextManager[sqlite3.Connection]:
+        connection = None
+        try:
+            connection = self.connect()
+            yield connection
+            connection.execute('COMMIT')
+            connection.close()
+
+        except Exception as e:
+            logger.exception(f'{e.__class__.__name__}: {e}')
+            if connection:
+                connection.close()  # pragma: no cover
+            if isinstance(e, sqlite3.DatabaseError):
+                self.filename.unlink(missing_ok=True)
+            raise
+
+    @with_retry
+    def atomic_get_primary_process(self, kind: ProcessKind,
+                                  current_process: Optional[TriblerProcess] = None) -> Optional[TriblerProcess]:
+        primary_process = None
+        with self.transaction() as connection:  # pylint: disable=not-context-manager  # false Pylint alarm
```

Comment:
NIT Why did you not use `transaction` as a variable name?
```suggestion
        with self.transaction() as transaction:  # pylint: disable=not-context-manager  # false Pylint alarm
```

Reply:
The new context manager is called `connect` instead of `transaction`, so it should be more intuitive now